### ISSUE TITLE
Architecture to handle User Functions

### ISSUE THREAD

**Issue by janvanrijn** [openml/openml.org#8]
Since the last OpenML workshop, @joaquinvanschoren and me have been discussing how to handle user functions such as:
* Authentication by username / password
* Registration of new user
* Reset password
* Update profile
* etc.. 

I thought it might be a good idea to create an issue, to have this discussion in the open with all stake holders. 

**The problem:** Since the start of OpenML, we use CodeIgniter in combination with `Ion_Auth`. In the beginning this was no problem, as all web-functionality was built in PHP, and other modules relied on a simplified authentication protocol based on a session hash. However, currently, the Python website also needs to be able to authenticate users by username and password. And there seem to be some slight problems. 

The options I see to solve this problem are the following:
* Port all functionality from PHP to Python. This basically would duplicate the relevant parts of `Ion_Auth` code to Python. Advantages are that we use only a small fraction of the code, so it should be fairly easy. Disadvantages are plenty: i) we would need to maintain this code ourselves. ii) Authentication is usually a fair bit harder than it seems, and there are many pitfalls. It is good to rely on third party software since I doubt we have this expertise ourselves. iii) it would make it hard to upgrade `Ion_Auth`, as this would also involve updates on the Python functions. 

* Expose the Ion_Auth functions to Python through a CLI interface. Advantages: shouldn't be as hard as it sounds. Disadvantages: probably also not as easy as it sounds

* Select a Python library that handles authentication, and expose the functions to PHP through a CLI-based interface. Same advantages and disadvantages as above. Additional pro: more future proof (as OpenML might in the distant future be completely Python based), Additional con: will be a very complicated procedure to migrate all data from the current db structure to the other one. 

WDYT?

**Comment by joaquinvanschoren** [comment]
TL;DR:
* Unless I'm missing something, there will be no need to use/maintain the PHP Ion_Auth anymore. This is an artifact of the old frontend being written in PHP. We can shift the functionality entirely to a modern Python library. The PHP backend then only needs to worry about running the OpenML API.
* The new frontend is truly 'standalone', it can run anywhere without having the OpenML PHP backend installed next to it. Hence, it shouldn't call a PHP CLI interface.

Long story:
This repo is about building a standalone frontend for OpenML instances. That means it should be able to work without having a local installation of OpenML running next to it. It's a simple Flask app that runs by itself, it only talks to the OpenML server through the REST API and ElasticSearch (it can have a user database, I'll get back to that).

Example use cases of standalone use:
* People building a scientific project on OpenML can set up their own frontend, filtering only the data and results relevant for them, and they can customize it as they like. 
* Someone wanting to improve the frontend can easily run it locally (no local OpenML install)

This is very lightweight, with a very small footprint. It is just a matter of 'git clone, run flask'. No local installations or database imports. (Unless you want to set up your own local OpenML instance, but that's a whole other use case). 

Now, we (obviously) also want to use this frontend as the frontend for the 'main' OpenML instance. This requires that the frontend has access to the OpenML user database. 

The most straightforward way to do this is to connect flask to the existing user database (rather than letting it create a new one) and use the 'built-in' modern user management libraries to handle everything described above. More info here: http://exploreflask.com/en/latest/users.html (we'll need to use argon2 instead of bcrypt though).
With this in place, the frontend will call on this library for all user management actions, and the old PHP Ion_Auth won't be called any more as far as I can see, since the old frontend will be replaced entirely?
The PHP backend will only be responsible for the API, where users authenticate with their api_keys.

**Comment by janvanrijn** [comment]
Sounds reasonable to have this main functionality in Python, as this should be the future of OpenML. 

However, after spending this week quite some time with ion auth, I feel more and more confident that these kind of things should not be programmed by hand, but rather using an out of the box library for this. 

What I read between the lines is that you currently plan to implement this functionality all by yourself, which incurs a heavy maintainability burden on us (please correct me if I am wrong).
